### ISSUE TITLE
how to gracefully shutdown a node

### ISSUE THREAD

**Issue by cameronbraid** [strimzi/strimzi-kafka-operator#878]
I had to restart a node and used the typical openshift procedure - mark node unschedulable, drain and reboot.

The kafka-kafka-1 pod remained after doing the drain operation, so I deleted it, assuming that it would trap the term signal and gracefully shutdown.

However the kafka pod came back with the following error logs 

```
2018-09-20 16:52:50,707 WARN [Log partition=request-event-4, dir=/var/lib/kafka/kafka-log1] Found a corrupted index file corresponding to log file /var/lib/kafka/kafka-log1/request-event-4/00000000000000000000.log due to Corrupt time index found, time index file (/var/lib/kafka/kafka-log1/request-event-4/00000000000000000000.timeindex) has non-zero size but the last timestamp is 0 which is less than the first timestamp 1537220023314}, recovering segment and rebuilding index files... (kafka.log.Log) [pool-6-thread-1]
```

The data loss is not a problem this time, however I would like to know how to avoid this in the future.

**Comment by cameronbraid** [comment]
I looked in the logs immediately after deleting it and I get the following 
```
2018-09-20 17:48:53,557 INFO Terminating process due to signal SIGTERM (org.apache.kafka.common.utils.LoggingSignalHandler) [SIGTERM handler]
2018-09-20 17:48:53,558 INFO Unable to read additional data from server sessionid 0x301473372870001, likely server has closed socket, closing socket connection and attempting reconnect (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:48:53,559 INFO [KafkaServer id=1] shutting down (kafka.server.KafkaServer) [kafka-shutdown-hook]
2018-09-20 17:48:53,560 INFO [KafkaServer id=1] Starting controlled shutdown (kafka.server.KafkaServer) [kafka-shutdown-hook]
2018-09-20 17:48:53,660 INFO [ZooKeeperClient] Waiting until connected. (kafka.zookeeper.ZooKeeperClient) [kafka-shutdown-hook]
2018-09-20 17:48:53,660 INFO [ZooKeeperClient] Connected. (kafka.zookeeper.ZooKeeperClient) [kafka-shutdown-hook]
2018-09-20 17:48:54,766 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:48:54,767 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:48:54,867 INFO [ZooKeeperClient] Waiting until connected. (kafka.zookeeper.ZooKeeperClient) [kafka-shutdown-hook]
2018-09-20 17:48:56,747 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:48:56,748 INFO Socket error occurred: localhost/127.0.0.1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:48:57,904 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:48:57,905 INFO Socket error occurred: localhost/127.0.0.1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:48:59,964 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:48:59,965 INFO Socket error occurred: localhost/127.0.0.1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:01,067 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:01,067 INFO Socket error occurred: localhost/127.0.0.1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:01,320 INFO Updated PartitionLeaderEpoch. New: {epoch:5, offset:299401}, Current: {epoch:4, offset:299180} for Partition: request-event-3. Cache now contains 4 entries. (kafka.server.epoch.LeaderEpochFileCache) [ReplicaFetcherThread-0-0]
2018-09-20 17:49:01,441 INFO Updated PartitionLeaderEpoch. New: {epoch:5, offset:298766}, Current: {epoch:4, offset:298557} for Partition: request-event-0. Cache now contains 4 entries. (kafka.server.epoch.LeaderEpochFileCache) [ReplicaFetcherThread-0-0]
2018-09-20 17:49:01,442 INFO Updated PartitionLeaderEpoch. New: {epoch:5, offset:300730}, Current: {epoch:4, offset:300548} for Partition: request-event-4. Cache now contains 4 entries. (kafka.server.epoch.LeaderEpochFileCache) [ReplicaFetcherThread-0-0]
2018-09-20 17:49:02,228 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:02,228 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:02,350 INFO Updated PartitionLeaderEpoch. New: {epoch:4, offset:298896}, Current: {epoch:3, offset:298676} for Partition: request-event-2. Cache now contains 3 entries. (kafka.server.epoch.LeaderEpochFileCache) [ReplicaFetcherThread-0-2]
2018-09-20 17:49:03,559 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:03,559 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:04,356 INFO Updated PartitionLeaderEpoch. New: {epoch:5, offset:299780}, Current: {epoch:4, offset:299565} for Partition: request-event-1. Cache now contains 4 entries. (kafka.server.epoch.LeaderEpochFileCache) [ReplicaFetcherThread-0-2]
2018-09-20 17:49:04,960 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:04,960 INFO Socket error occurred: localhost/127.0.0.1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:06,099 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:06,099 INFO Socket error occurred: localhost/127.0.0.1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:07,857 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:07,857 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:09,428 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:09,428 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:11,307 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:11,307 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:13,004 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:13,004 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:14,151 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:14,152 INFO Socket error occurred: localhost/127.0.0.1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:15,788 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:15,789 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:17,244 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:17,245 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:18,630 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:18,630 INFO Socket error occurred: localhost/127.0.0.1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:19,626 INFO Updated PartitionLeaderEpoch. New: {epoch:4, offset:6395}, Current: {epoch:2, offset:6373} for Partition: turbo-search-results-view-4. Cache now contains 2 entries. (kafka.server.epoch.LeaderEpochFileCache) [ReplicaFetcherThread-0-2]
2018-09-20 17:49:20,506 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:20,506 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:22,382 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2018-09-20 17:49:22,382 INFO Socket error occurred: localhost/127.0.0.1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
```

Is there something mis-configured ?

**Comment by scholzj** [comment]
What is your configuration? How many Zookeeper nodes do you have? AFAIK SIGTERM is the usual way how Kafka stops it self (see [kafka-server-stop.sh](https://github.com/apache/kafka/blob/trunk/bin/kafka-server-stop.sh)). The problem from your log seems to be that the broker lost connection to Zoo and that caused that it didn't shutdown properly. I'm not on 100% sure whether this is even a bug - because it might need the Zookeeper to coordinate.

That is why I'm wondering how many Zookeeper nodes did you had and whether they shared the same host.

**Comment by cameronbraid** [comment]
Running on Openshift Origin  3.,7

Pretry straight forward, using local-provisioned ssd storage
*  3 Zookeeper replicas one on node01, one on node02 and one on node03
* 3 Kafka replicas one on node01, and one onnode02 and and one onnode03

```
apiVersion: kafka.strimzi.io/v1alpha1
kind: Kafka
metadata:
  clusterName: ''
  creationTimestamp: '2018-09-10T01:21:25Z'
  deletionGracePeriodSeconds: null
  deletionTimestamp: null
  name: kafka
  namespace: drivenow-staging-x
  resourceVersion: '44141550'
  selfLink: /apis/kafka.strimzi.io/v1alpha1/namespaces/drivenow-staging-x/kafkas/kafka
  uid: d6163eb9-b497-11e8-aaae-0cc47a39144c
spec:
  entityOperator:
    topicOperator: {}
    userOperator: {}
  kafka:
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.min.isr: 2
      transaction.state.log.replication.factor: 3
    listeners:
      plain: {}
      tls: {}
    livenessProbe:
      initialDelaySeconds: 15
      timeoutSeconds: 5
    metrics:
      lowercaseOutputName: true
      rules:
        - name: kafka_server_$1_$2_total
          pattern: 'kafka.server<type=(.+), name=(.+)PerSec\w*><>Count'
        - labels:
            topic: $3
          name: kafka_server_$1_$2_total
          pattern: 'kafka.server<type=(.+), name=(.+)PerSec\w*, topic=(.+)><>Count'
    readinessProbe:
      initialDelaySeconds: 15
      timeoutSeconds: 5
    replicas: 3
    storage:
      class: local-ssd
      deleteClaim: false
      size: 1Gi
      type: persistent-claim
  zookeeper:
    livenessProbe:
      initialDelaySeconds: 15
      timeoutSeconds: 5
    metrics:
      lowercaseOutputName: true
    readinessProbe:
      initialDelaySeconds: 15
      timeoutSeconds: 5
    replicas: 3
    storage:
      class: local-ssd
      deleteClaim: false
      size: 1Gi
      type: persistent-claim
```

**Comment by scholzj** [comment]
Hmm, I guess the problem might not be Zookeeper it self but the TLS sidecar whcih maybe shutdown before Kafka. I will have a look if we can do something about it.

**Comment by cameronbraid** [comment]
Yeah, good point about the TLS sidecar, that could be a likely cause.  

One thing that I don't understand, that since only 1 kafka pod shutdown, there are two others, and the replication factor for all my topics is 3, wouldn't the same data be committed on the other 2 nodes so there should not have been any data loss (apart from any un-acked writes), even if it was kill-9-ed

**Comment by cameronbraid** [comment]
For example one topic went from 1g on (according to kafka_server_brokertopicmetrics_bytesin_total) to 670mb

**Comment by scholzj** [comment]
So, to be honest, I do not think the message about rebuilding the index on its own necessarily means a data loss. I think it more or less only means that the index file was not closed cleanly and it will be rebuilt.

The replication factor 3 will not guarantee no data lost on its own. To make sure there is no data loss, you need to make sure that:
* You always have at least 2 in-sync replicas (`min.insync.replicas`)
* You need to set the `acks` option in producer to all, to make sure that producer sees the message as delivered only when it is written to all in-sync replicas.
Without these it can always happen that the message will be written only to the leader and lost when the leader dies immediately afterwards.

The `kafka_server_brokertopicmetrics_bytesin_total` metric is AFAIK a counter on received bytes. So when the broker restarts, the counter restarts IMHO as well. So I do not think this indicates you lost 300MB of messages.

**Comment by cameronbraid** [comment]
Thanks for your quick reply and that all makes sense.  I appreciate the help and clarification.

**Comment by scholzj** [comment]
The issue with the TLS sidecar shutting down before the Kafka container should be addressed in PR #1190
